Source/Core/AbsyExpr.cs

```diff
@@ -3837,6 +3843,197 @@ public T Dispatch<T>(IAppliableVisitor<T> visitor)
+    void ObjectInvariant()
+    {
+      Contract.Invariant(this._tok != null);
+    }
+
+    public FieldAccess(IToken tok, string fieldName)
+    {
+      Contract.Requires(tok != null);
+      this._tok = tok;
+      this._fieldName = fieldName;
+    }
+    
+    public FieldAccess(IToken tok, DatatypeSelector selector)
+    {
+      Contract.Requires(tok != null);
+      this._tok = tok;
+      this._selector = selector;
+      this._fieldName = selector.Name;
+    }
+
+    public string FunctionName
+    {
+      get
+      {
+        Contract.Ensures(Contract.Result<string>() != null);
+
+        return "field-access";
+      }
+    }
+
+    [Pure]
+    [Reads(ReadsAttribute.Reads.Nothing)]
+    public override bool Equals(object obj)
+    {
+      if (obj is FieldAccess fieldAccess)
+      {
+        return selector.Equals(fieldAccess.selector);
+      }
+      return false;
+    }
+
+    [Pure]
+    public override int GetHashCode()
+    {
+      return 1;
+    }
+
+    public void Emit(IList<Expr> args, TokenTextWriter stream, int contextBindingStrength, bool fragileContext)
+    {
+      stream.SetToken(this);
+      Contract.Assert(args.Count == 1);
+      stream.push();
+      stream.Write("(");
+      cce.NonNull(args[0]).Emit(stream, 0x00, false);
+      stream.Write("->{0}", fieldName);
+      stream.Write(")");
+      stream.pop();
+    }
+
+    public void Resolve(ResolutionContext rc, Expr subjectForErrorReporting)
+    {
+      // The work of resolution is delayed to type checking when the datatype is known.
+    }
+
+    public int ArgumentCount
+    {
+      get { return 1; }
+    }
+
+    public Type Typecheck(IList<Expr> args, out TypeParamInstantiation tpInstantiation, TypecheckingContext tc)
+    {
+      Contract.Ensures(args != null);
+      Contract.Ensures(Contract.ValueAtReturn(out tpInstantiation) != null);
+      Contract.Assert(args.Count == 1);
+      return Typecheck(cce.NonNull(args[0]).Type, tc, out tpInstantiation);
+    }
+    
+    public Type Typecheck(Type type, TypecheckingContext tc, out TypeParamInstantiation tpInstantiation)
+    {
+      tpInstantiation = SimpleTypeParamInstantiation.EMPTY;
+      if (!(type is CtorType ctorType))
+      {
+        tc.Error(this.tok, "type {0} is not a constructor type", type);
+        return null;
+      }
+      if (!(ctorType.Decl is DatatypeTypeCtorDecl datatypeTypeCtorDecl))
+      {
+        tc.Error(this.tok, "field-access must be applied to a datatype, {0} is not a datatype", ctorType);
+        return null;
+      }
+      var selectors = datatypeTypeCtorDecl.GetSelectors(fieldName);
+      if (selectors == null)
+      {
+        tc.Error(this.tok, "datatype {0} does not have a field with name {1}", ctorType, fieldName);
+        return null;
+      }
+      Contract.Assert(selectors.Count > 0);
+      if (selectors.Count > 1)
+      {
+        tc.Error(this.tok, "datatype {0} has several fields with name {1}", ctorType, fieldName);
```

Comment:
Just as a point of comparison, Dafny allows a datatype to have several destructors with the same name, as long as they all have the same type. No need to change anything at this time, but if this further functionality is needed in the future, it could be supported in Boogie.

Also, in Dafny, expressions can be partial, so a precondition of such a field-access operation on a datatype is that the variant (constructor) of the datatype is one that has the requested field. Since Boogie is total, I imagine that Boogie always allows the field access, but sometimes acts like just an uninterpreted function.

Reply:
The second paragraph in your comment above is accurate.